[16.0][MIG] shift_worker_status

remytms opened the pull request:
## Description

Migration of module shift_worker_status

## Odoo task (if applicable)

[task](https://gestion.coopiteasy.be/web#id=12148&model=project.task&view_type=form&menu_id=536)

## Checklist before approval

- [x] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [x] Change log snippet is present.
- [x] (If a new module) Moving this to OCA has been considered.

carmenbianca committed:
[REF] Rename module: beesdoo_worker_status -> shift_worker_status

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift_worker_status: Add OCA to authors

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] shift_worker_status: Pre-init hook

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] shift_worker_status: beesdoo_shift -> shift

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] shift_worker_status: beesdoo.shift -> shift

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] shift_worker_status: beesdoo_worker_status -> shift_worker_status

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift_worker_status: Manually remove beesdoo references

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] shift_worker_status: task -> shift_shift

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] Satisfy pre-commit

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk committed:
[FIX] fix beesdoo renaming scripts

* rename many2many tables and their columns.
* rename constraints and indexes.

carmenbianca committed:
[FIX] shift_worker_status: Copy old value from parameter

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift: Prefix parameters in code

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] Add logging entry

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] Actually preserve old param values

There are two fixes in this commit:

- Use .search() instead of .get_param() for a more robust search of the
  old value.
- Rename all XMLIDs. When they XMLIDs aren't renamed, the values get
  overridden during init, even though the config_parameter keys were
  renamed.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift_worker_status: Rename file

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift_worker_status: rename class

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift_worker_status: rename test class

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] Run pre-commit

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] shift_worker_status: rename test file

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] author fields in manifests

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[UPD] bump versions

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

Commit:
[UPD] Update shift_worker_status.pot

github-grap-bot committed:
[UPD] README.rst

huguesdk committed:
[FIX] fix penalty setting resetting on update

fix shift_worker_status.irregular_penalty config parameter resetting to
true when updating the module.

github-grap-bot committed:
shift_worker_status 12.0.2.0.1

github-grap-bot committed:
[BOT] post-merge updates

remytms committed:
[MIG] shift_worker_status: migration to 16.0

remytms committed:
[FIX] shift_worker_status: tests

remytms committed:
[FIX] shift_worker_status: state change hook

_write function is not called the same way than in 12.0.

The _state_change() hook is therefore not triggered correctly.

_state_change() is now added to the _compute_state function as for the
log in status.history.

remytms committed:
[IMP] shift_worker_status: tests

polchampion reviewed (approved):
functional test on 16-test-shift
@remytms can you add the features listed below in the DESCRIPTION file? 
see also my comments below fyi

**features**
- [x] add "irregular penalty" config in shift settings
- [x] compute next countdown date for irregular workers, taking into account holidays and exemption status
- [x] compute the future alert date for irregular workers but without taking holidays into account (to add to roadmap)
- [x] compute regular and irregular statuses according to alert start day, alert delay, grace delay and time extensions, taking into account holidays and exemption status
- [x] recompute status and counters when a shift undergoes a state change
- [x] postpone alert start time based on holidays and extension status
- [x] handle is_regular and is_compensation consistency on shifts, on worker change and by raising a validation error if the fields are not properly set

**some remarks¨**
- contrary to v12, it's not possible to change the attendance state of a shift by clicking on the state names on the top right in form view. 
- "Future Alert Date" and "Next Countdown Date" are computed according to a hardcoded 28-day period instead of depending on "shift.shift_period"
- The next countdown date takes into account holidays and exemptions, but sadly the future alert date does not...
- when shift_worker_status isn't installed, there is an error (see below) preventing to subscribe a worker to a working mode

```RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/home/odoo16/src/odoo/odoo/http.py", line 1652, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/home/odoo16/src/odoo/odoo/service/model.py", line 133, in retrying
    result = func()
  File "/home/odoo16/src/odoo/odoo/http.py", line 1679, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/home/odoo16/src/odoo/odoo/http.py", line 1883, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/home/odoo16/src/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/home/odoo16/src/odoo/odoo/http.py", line 734, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/home/odoo16/src/odoo/addons/web/controllers/dataset.py", line 46, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/home/odoo16/src/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/odoo16/src/odoo/odoo/api.py", line 469, in call_kw
    model.env.flush_all()
  File "/home/odoo16/src/odoo/odoo/api.py", line 755, in flush_all
    self._recompute_all()
  File "/home/odoo16/src/odoo/odoo/api.py", line 751, in _recompute_all
    self[field.model_name]._recompute_field(field)
  File "/home/odoo16/src/odoo/odoo/models.py", line 6283, in _recompute_field
    field.recompute(records)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1370, in recompute
    apply_except_missing(self.compute_value, recs)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1343, in apply_except_missing
    func(records)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1392, in compute_value
    records._compute_field_value(self)
  File "/home/odoo16/src/odoo/addons/mail/models/mail_thread.py", line 403, in _compute_field_value
    return super()._compute_field_value(field)
  File "/home/odoo16/src/odoo/odoo/models.py", line 4241, in _compute_field_value
    fields.determine(field.compute, self)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 101, in determine
    return needle(records, *args)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 702, in _compute_related
    record[self.name] = self._process_related(value[self.related_field.name])
  File "/home/odoo16/src/odoo/odoo/models.py", line 5957, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1158, in __get__
    self.recompute(record)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1370, in recompute
    apply_except_missing(self.compute_value, recs)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1343, in apply_except_missing
    func(records)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 1392, in compute_value
    records._compute_field_value(self)
  File "/home/odoo16/src/odoo/odoo/models.py", line 4241, in _compute_field_value
    fields.determine(field.compute, self)
  File "/home/odoo16/src/odoo/odoo/fields.py", line 98, in determine
    return needle(*args)
  File "/home/odoo16/src/coopiteasy/obeesdoo/shift/models/cooperative_status.py", line 183, in _compute_status
    rec._state_change(previous_status, new_status)
TypeError: CooperativeStatus._state_change() takes 2 positional arguments but 3 were given

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    RPCError@https://odoo16.main.test.srv.coopiteasy.be/web/assets/227-3c0999b/web.assets_backend.min.js:993:274
    makeErrorFromResponse@https://odoo16.main.test.srv.coopiteasy.be/web/assets/227-3c0999b/web.assets_backend.min.js:997:163
    jsonrpc/promise</<@https://odoo16.main.test.srv.coopiteasy.be/web/assets/227-3c0999b/web.assets_backend.min.js:1005:34
    ```

remytms committed:
[FIX] shift: statusbar clickable

remytms committed:
[FIX] shift_worker_status: compute dates

next_countdown_date and future_alert_date were not computed the same
way, leading to differences for an irregular worker, with a counter (sr)
equal to zero and an exemption set during the next countdown date.

polchampion commented:
all good

**features**
- [x] add "irregular penalty" config in shift settings
- [x] compute next countdown date and future alert date for irregular workers, taking into account holidays and exemption status
- [x] compute the future alert date for irregular workers but without taking holidays into account (to add to roadmap)
- [x] compute regular and irregular statuses according to alert start day, alert delay, grace delay and time extensions, taking into account holidays and exemption status
- [x] recompute status and counters when a shift undergoes a state change
- [x] postpone alert start time based on holidays and extension status
- [x] handle is_regular and is_compensation consistency on shifts, on worker change and by raising a validation error if the fields are not properly set

polchampion commented:
@remytms Good work! can you add the above features list to the module description and then merge?

remytms committed:
[IMP] shift_worker_status: readme

remytms commented:
/ocabot merge nobump
